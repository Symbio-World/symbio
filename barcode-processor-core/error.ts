export const NoSearchResultsFound = 'NoSearchResultsFound'
export type NoSearchResultsFound = typeof NoSearchResultsFound
export const SearchBarcodeRequestFailed = 'SearchBarcodeRequestFailed'
export type SearchBarcodeRequestFailed = typeof SearchBarcodeRequestFailed
export type SearchBarcodeError =
  | NoSearchResultsFound
  | SearchBarcodeRequestFailed
export type ProcessBarcodeError = SearchBarcodeError
