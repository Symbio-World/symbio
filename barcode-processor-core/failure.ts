import { F } from '@symbio/ts-lib'

export const NO_SEARCH_RESULTS_FOUND = F.createFailure('NO_SEARCH_RESULTS_FOUND')
export type NO_SEARCH_RESULTS_FOUND = typeof NO_SEARCH_RESULTS_FOUND

export type SearchBarcodeFailure =
  | F.HTTP_FAILURE
  | F.DECODING_FAILURE
  | NO_SEARCH_RESULTS_FOUND

export type TranslateProductDataFailure =
  | F.HTTP_FAILURE
  | F.DECODING_FAILURE

export type FetchProductPageFailure =
  | F.HTTP_FAILURE
  | F.DECODING_FAILURE

export type ProcessBarcodeFailure =
  | SearchBarcodeFailure
  | TranslateProductDataFailure
  | FetchProductPageFailure
