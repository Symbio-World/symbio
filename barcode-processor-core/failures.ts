import { createFailure } from '@symbio/ts-lib'
import { ProductData } from './ProductData'

export const NO_SEARCH_RESULTS_FOUND = 'NO_SEARCH_RESULTS_FOUND'
export type NO_SEARCH_RESULTS_FOUND = typeof NO_SEARCH_RESULTS_FOUND
export const noSearchResultsFound = (barcode: string) =>
  createFailure(NO_SEARCH_RESULTS_FOUND)(
    `Found no items for barcode ${barcode}`,
  )

export const NO_USEFUL_INFO_FOUND = 'NO_USEFUL_INFO_FOUND'
export type NO_USEFUL_INFO_FOUND = typeof NO_USEFUL_INFO_FOUND
export const noUsefulInfoFound = (barcode: string, productData: ProductData) =>
  createFailure(NO_USEFUL_INFO_FOUND)(
    `No useful information found for barcode ${barcode}. The information found is as follows: ${JSON.stringify(productData)}`,
  )
