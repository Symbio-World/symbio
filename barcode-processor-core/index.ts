export * from './ProductData'
export * from './error'
export * from './createProcessBarcode'
