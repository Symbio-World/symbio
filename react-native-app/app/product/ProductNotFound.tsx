import * as React from 'react'
import { View, Text, Linking } from 'react-native'
import { t } from 'react-native-tailwindcss'
import { Button } from '../ui-kit/Button'
import { Modal } from '../ui-kit/Modal'

type Props = {
  barcode?: string
}
export const ProductNotFound: React.FC<Props> = ({ barcode }) => {
  const [isPressed, setPressed] = React.useState<boolean>(false)
  return (
    <View style={[t.flex1, t.justifyCenter, t.m5]}>
      {isPressed && (
        <Modal onDismiss={() => setPressed(false)}>
          <View style={[t.flex1, t.justifyCenter, t.m5]}>
            <Text>Thanks for helping, this feature will be available soon</Text>
          </View>
        </Modal>
      )}
      <Text>
        Unfortunately we could not find any data for this barcode{' '}
        <Text>{barcode}</Text>
      </Text>
      <View style={[t.mY5]}>
        <Button
          title="Search with google"
          onPress={() =>
            Linking.openURL(`https://www.google.com/search?q=${barcode}`)
          }
        />
      </View>
      <Button title="Add Manually" onPress={() => setPressed(true)} />
    </View>
  )
}
