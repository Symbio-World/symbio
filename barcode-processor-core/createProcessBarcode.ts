import * as R from 'ramda'
import { allSettled, PromiseResolution } from '@symbio/ts-lib'
import * as Model from './ProductData'
import { noUsefulInfoFound } from './failures'

export type SearchBarcode = (
  barcode: string,
) => Promise<Model.ProductSearchData>

export type FetchProductPage = (link: string) => Promise<Model.ProductPage>

export type ScrapeProductPage = (
  productPage: Model.ProductPage,
) => Model.ProductPageData

export type TranslateProductData = (
  productData: Model.ProductData,
) => Promise<Model.ProductData>

export type ProcessBarcode = (b: string) => Promise<Model.ProductData>

export type Deps = {
  searchBarcode: SearchBarcode
  fetchProductPage: FetchProductPage
  scrapeProductPage: ScrapeProductPage
  translateProductData: TranslateProductData
}
export type CreateProcessBarcode = (deps: Deps) => ProcessBarcode
export const createProcessBarcode: CreateProcessBarcode = ({
  searchBarcode,
  fetchProductPage,
  scrapeProductPage,
  translateProductData,
}) => async (barcode) => {
  console.log('initiating barcode search...')
  const productSearchData = await searchBarcode(barcode)
  console.log(
    `barcode search completed with ${JSON.stringify(
      productSearchData,
      null,
      4,
    )}`,
  )
  const productPageRequests = productSearchData.links.map(fetchProductPage)
  console.log(`initiating requests to product pages...`)
  const responses = await allSettled(productPageRequests)
  console.log(`received ${responses.length} responses`)
  const productPages = responses
    .filter((result) => result.status === 'fulfilled')
    .map((result) => (result as PromiseResolution<Model.ProductPage>).value)
  console.log(
    `starting to scrape data from ${productPages.length} product pages`,
  )
  const dataFromProductPages = productPages.map(scrapeProductPage)
  console.log(`data scraped: ${JSON.stringify(dataFromProductPages, null, 4)}`)
  const combinedProductPageData = dataFromProductPages.reduce(
    (acc, pageData) => R.mergeDeepLeft(acc, pageData) as Model.ProductPageData,
    {},
  )
  const productData = {
    ...productSearchData,
    ...combinedProductPageData,
  }

  console.log(
    `product data before translation: ${JSON.stringify(productData, null, 4)}`,
  )

  if (R.isEmpty(R.omit(['links'], productData))) {
    throw noUsefulInfoFound(barcode, productData)
  }

  const translatedProductData = await translateProductData(productData)
  console.log(
    `translated product data: ${JSON.stringify(
      translatedProductData,
      null,
      4,
    )}`,
  )
  return translatedProductData
}
