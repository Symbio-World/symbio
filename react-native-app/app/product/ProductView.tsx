import * as React from 'react'
import { View, Image, Text, ScrollView } from 'react-native'
import { t } from 'react-native-tailwindcss'
import { ProductData } from '@symbio/barcode-processor-core'
import { ModalActionsView } from '../ui-kit/ModalActionsView'
import { CloseButton } from '../ui-kit/CloseButton'

type Props = Omit<ProductData, 'links'> & {
  onClose?: () => void
}

export const ProductView: React.FC<Props> = ({
  image = '',
  name = '',
  description = '',
  ingredients = '',
  allergens = [],
  origin = '',
  onClose,
}) => {
  return (
    <View style={[t.flex1, t.mB3]}>
      <ScrollView style={[t.bgWhite, t.flex1]}>
        <Image
          style={[t.h56, t.bgGray200]}
          source={{ uri: image }}
          resizeMode="contain"
        />
        <View style={[t.p8]}>
          <Text style={[t.textLg, t.mB8]}>{name}</Text>
          <Text>{description}</Text>
          <Text style={[t.fontBold, t.mT8]}>Ingredients</Text>
          <Text>{ingredients ? ingredients : '-'}</Text>
          <Text style={[t.fontBold, t.mT8]}>Allergens</Text>
          {allergens.map((a) => (
            <Text>{a}</Text>
          ))}
          <Text style={[t.fontBold, t.mT8]}>Origin</Text>
          <Text>{origin}</Text>
        </View>
      </ScrollView>
      <ModalActionsView>
        <CloseButton onClose={onClose} />
      </ModalActionsView>
    </View>
  )
}
