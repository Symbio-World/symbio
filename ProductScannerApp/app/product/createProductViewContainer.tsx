import React from 'react'
import useSWR from 'swr'
import * as R from 'ramda'
import { ProductData, ProcessBarcode } from '@symbio/barcode-processor-core'
import { Loading } from '../ui-kit/Loading'
import { Error } from '../ui-kit/Error'
import { ProductView } from './ProductView'
import { ProductNotFound } from './ProductNotFound'

type Props = {
  barcode: string
}

type Deps = {
  processBarcode: ProcessBarcode
}

type CreateProductViewContainer = (deps: Deps) => React.FC<Props>

export const createProductViewContainer: CreateProductViewContainer = ({
  processBarcode,
}) => ({ barcode }) => {
  const { data: productData, isValidating: isLoading, error } = useSWR<
    ProductData
  >(barcode, processBarcode)

  if (isLoading) return <Loading />
  if (error) return <Error error={error} />
  if (!productData || R.isEmpty(productData))
    return <ProductNotFound barcode={barcode} />
  return <ProductView {...productData} />
}
