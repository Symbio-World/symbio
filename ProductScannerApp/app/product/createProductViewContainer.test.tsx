import React from 'react'
import { render } from 'react-native-testing-library'
import { ActivityIndicator } from 'react-native'
import * as Core from '@symbio/barcode-processor-core'
import { createProductViewContainer } from './createProductViewContainer'
import { ProductNotFound } from './ProductNotFound'
import { ProductView } from './ProductView'

describe('createProductViewContainer', () => {
  it('renders correctly', () => {
    const ProductViewContainer = createProductViewContainer({
      processBarcode: jest.fn(),
    })
    const { toJSON } = render(<ProductViewContainer barcode="6414893012318" />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('renders loading screen at the start', () => {
    const ProductViewContainer = createProductViewContainer({
      processBarcode: jest.fn(),
    })
    const { getByType } = render(
      <ProductViewContainer barcode="6414893012318" />,
    )
    expect(getByType(ActivityIndicator)).toBeDefined()
  })

  it('calls process barcode', () => {
    const promise = Promise.resolve({ links: [] })
    const processBarcode: Core.ProcessBarcode = jest.fn(() => promise)
    const ProductViewContainer = createProductViewContainer({
      processBarcode,
    })
    const barcode = '6414893012311'
    const productViewContainer = <ProductViewContainer barcode={barcode} />
    render(productViewContainer)
    expect(processBarcode).toHaveBeenCalledWith(barcode)
  })

  it('renders no product found if product is empty', async () => {
    const promise = Promise.resolve({} as Core.ProductData)
    const processBarcode: Core.ProcessBarcode = jest.fn(() => promise)
    const ProductViewContainer = createProductViewContainer({
      processBarcode,
    })
    const productViewContainer = (
      <ProductViewContainer barcode="6414893012312" />
    )
    const { getByType } = render(productViewContainer)
    await promise
    expect(getByType(ProductNotFound)).toBeDefined()
  })

  it('renders product card if data fetch succeeded', async () => {
    const product = { name: 'Margarin', links: [] }
    const promise = Promise.resolve(product)
    const processBarcode: Core.ProcessBarcode = jest.fn(() => promise)
    const ProductViewContainer = createProductViewContainer({
      processBarcode,
    })
    const productViewContainer = (
      <ProductViewContainer barcode="6414893012313" />
    )
    const { getByType } = render(productViewContainer)
    await promise
    expect(getByType(ProductView)).toBeDefined()
  })
})
