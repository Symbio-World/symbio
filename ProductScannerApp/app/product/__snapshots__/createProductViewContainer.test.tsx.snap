// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProductViewContainer renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#48bb78"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
